src/NovelRT/Experimental/Graphics/GraphicsMemoryBlockCollection.cpp

```diff
@@ -0,0 +1,481 @@
+
+                if ((budget.GetEstimatedUsage() >= budget.GetEstimatedBudget()) &&
+                    (blocksCount > static_cast<size_t>(GetMinimumBlockCount())) &&
+                    ((GetSize() - block->GetSize()) >= GetMinimumSize()))
+                {
+                    RemoveBlockAt(blockIndex);
+                }
+                else
+                {
+                    _emptyBlock = block;
+                }
+            }
+        }
+
+        IncrementallySortBlocks();
+    }
+
+    bool GraphicsMemoryBlockCollection::TryAllocate(size_t size,
+                                                    size_t alignment,
+                                                    GraphicsMemoryRegionAllocationFlags flags,
+                                                    GraphicsMemoryRegion<GraphicsMemoryBlock>& outRegion)
+    {
+        std::lock_guard<std::mutex> guard(_mutex); // TODO: come back to this when Tanner gets back to you
+        return TryAllocateRegion(size, alignment, flags, outRegion);
+    }
+
+    bool GraphicsMemoryBlockCollection::TryAllocate(size_t size, GraphicsMemoryRegion<GraphicsMemoryBlock>& outRegion)
+    {
+        return TryAllocate(size, 1, GraphicsMemoryRegionAllocationFlags::None, outRegion);
+    }
+
+    bool GraphicsMemoryBlockCollection::TryAllocate(size_t size,
+                                                    size_t alignment,
+                                                    GraphicsMemoryRegionAllocationFlags flags,
+                                                    gsl::span<GraphicsMemoryRegion<GraphicsMemoryBlock>> regions)
+    {
+        bool succeeded = true;
+        size_t index;
+
+        // This scope is needed for the lock guard.
+        {
+            std::lock_guard<std::mutex> guard(_mutex); // TODO: come back to this when Tanner gets back to you
+
+            for (index = 0; index < regions.size(); ++index)
+            {
+                succeeded = TryAllocateRegion(size, alignment, flags, regions[index]);
+
+                if (!succeeded)
+                {
+                    break;
+                }
+            }
+        }
+
+        if (!succeeded)
+        {
+            while (index-- != 0)
+            {
+                Free(regions[index]);
+                regions[index] = GraphicsMemoryRegion<GraphicsMemoryBlock>();
+            }
+        }
+
+        return succeeded;
+    }
+
+    bool GraphicsMemoryBlockCollection::TryAllocate(size_t size,
+                                                    gsl::span<GraphicsMemoryRegion<GraphicsMemoryBlock>> regions)
+    {
+        return TryAllocate(size, 1, GraphicsMemoryRegionAllocationFlags::None, regions);
+    }
+
+    bool GraphicsMemoryBlockCollection::TrySetMinimumSize(size_t minimumSize)
+    {
+        std::lock_guard<std::mutex> guard(_mutex); // TODO: come back to this when Tanner gets back to you
```

Comment:
Gonna assume the same here -_Tannnneerrrrrrrrrrrrrrrr_?

Reply:
Yeah I had a discussion with him, but I think this ultimately boiled down to "leave it alone for now".